rfcs/0042-bitswap.md

```diff
@@ -0,0 +1,268 @@
+
+```ocaml
+let open Core in
+let open Mina_transition in
+let precomputed_block = External_transition.Precomputed_block.t_of_sexp @@ Sexp.of_string External_transition_sample_precomputed_block.sample_block_sexp in
+let small_precomputed_block = {precomputed_block with staged_ledger_diff = Staged_ledger_diff.empty_diff} in
+let conv (t : External_transition.Precomputed_block.t) =
+  External_transition.create
+    ~protocol_state:t.protocol_state
+    ~protocol_state_proof:t.protocol_state_proof
+    ~staged_ledger_diff:t.staged_ledger_diff
+    ~delta_transition_chain_proof:t.delta_transition_chain_proof
+    ~validation_callback:(Mina_net2.Validation_callback.create_without_expiration ())
+    ()
+in
+Protocol_version.set_current (Protocol_version.create_exn ~major:0 ~minor:0 ~patch:0) ;
+External_transition.Stable.Latest.bin_size_t (conv small_precomputed_block) ;;
+```
+
+### Shipping as a Soft-Fork
+
+We have an option to ship Bitswap as a soft-fork upgrade rather than as a hard-fork upgrade (shoutout to @mrmr1993 for this suggestion; writing this nearly verbatim from him). We can do this by adding a new pub/sub topic for broadcasting the new block header broadcast message format, which we can support in parallel to the current pub/sub topic we use for all of our broadcast messages. We currently use the `"coda/consensus-messages/0.0.1"` topic for blocks, transactions, and snark work. Nodes running the new version of the software can still support this topic, but in addition, can subscribe to and interact over a new topic `"mina/blocks/1.0.0"`, where we can broadcast the new message format for blocks. Nodes are able to filter subscriptions from other nodes based on what they subscribe to, configured using the [`WithSubscriptionFilter` option](https://github.com/libp2p/go-libp2p-pubsub/blob/55d412efa7f5a734d2f926e0c7c948f0ab4def21/subscription_filter.go#L36). Utilizing this, nodes that support the `"mina/blocks/1.0.0"` can filter out the `"coda/consensus-messages/0.0.1"` topic from nodes that support both topics. By filtering the topics like this, nodes running the new version can broadcast new blocks over both topics while avoiding sending the old message format to other nodes which support the new topic. Then, in the next hard fork, we can completely deprecate sending blocks over the old topic.
+
+The main detail to still figure out here is how new nodes will bridge the new topic messages to the old topic. It may be as simple as just broadcasting the block as a fresh message on the old topic, but that isn't normally how a pub/sub rebroadcast cycle would operate. This approach may "just work", but we should do additional research into what differentiates a libp2p pub/sub broadcast from a rebroadcast. If it turns out the only difference is how we select which peers to send the message to, the proposed topic filtering solution may address it automatically.
```

Comment:
if I understand this right, this is the case where a block is received over the new topic, and we would want to re-gossip it over both the new and old topics to make sure both upgraded and non-upgraded peers get it? I think if the original broadcaster of the block gossips it over both topics, it *should* get propagated over both topics, so nodes receiving the block wouldn't need to re-gossip it over both topics

Reply:
Would this be true even if only a small number of nodes on the network were subscribing to the old topic? I was worried that, should we not have new nodes bridge the topic gap, that once the number of nodes running the old client gets so small, broadcasts over that topic might not be consistent. But perhaps this is a misunderstanding of how gossip pub/sub traffic is implemented in libp2p.

Reply:
My intention was to always broadcast over both, so then any peer of an 'old-gossip-only' node that receives such a gossip will re-broadcast it to them over the old channel. Concretely:
* we have to listen on the old-gossip topic to receive blocks from nodes that haven't upgraded
* we don't want to receive any messages over the old topic from any peers who support the new topic
  - part of the intention of this is supposed to be reducing bandwidth usage, so it doesn't make sense here for two peers connected over the new topic to ever communicate over the old topic, else they're both double-handling blocks and not reducing network traffic
* by the time we gossip a block, we've already validated it, so we can be confident sending it over the old topic signed by our libp2p key instead of the originator's